### DISCUSSION TITLE
How does `use_signal` guarantees memory safety?

### DISCUSSION THREAD

**Discussion by NobodyXu** [DioxusLabs/dioxus#2261] [General]
I'm reading documentation on hooks and [`dioxus-signals`](https://docs.rs/dioxus-signals).

I'm a bit confused on how `use_signal` upholds memory safety.
It seems that the signal returns from `use_signal` can be used in `spawn` (or `tokio::spawn` which requires `Sync`), which could run the task in a separate thread concurrently.

Based on the API of `use_signal` return type, I suspect it works similar to a `Cell`, where elements must be cloned when access to avoid holding the lock, but it would still need a lock anyway.

How does it handle that without `unsafe` code?
Is it because that on every platform, dioxus is compiled to wasm and thus has only one thread for UI?

**Comment by ealmloff** [comment]
Signal has a storage generic that controls if the signal is `Send + Sync`. The unsync version uses a `RefCell` and the sync version uses a `RwLock`. The read and write methods hold a lock. Calling the signal like a function just reads and clones the value

**Reply by NobodyXu** [reply]
Thank you!

> Signal has a storage generic that controls if the signal is `Send + Sync`.

So if I use signal in a `spawn` it will automatically use the sync version, thus wrap it in a `RwLock`?

> The unsync version uses a `RefCell` and the sync version uses a `RwLock`. The read and write methods hold a lock.

Looks like state management in dioxus is a bit more expensive than React in JS.

AFAIK wasm in browser cannot spawn thread (and thus no atomic), so does `RefCell` and `RwLock` translates to simple arithmetic of integers?

Though I'm not sure how blocking of locker is implemented.

> Calling the signal like a function just reads and clones the value

Does the jsx clones the value, or does it just read it?

**Reply by ealmloff** [reply]
In dioxus, spawn is single threaded. If you need a sync signal for a tokio task, you can `use_sync_signal`.

Dioxus (and WASM) doesn't have the overhead of garbage collection, and it uses templates to accelerate diffing/rendering. Overall Dioxus [tends to be faster than React](https://krausest.github.io/js-framework-benchmark/2024/table_chrome_123.0.6312.59.html), but I don't have benchmarks that specifically compare state management performance 

Targeting dioxus-web, the sync and unsync version should compile to similar code because atomics don't exist in wasm.

Using the value in a formatted string (`"{signal}"`) just reads the value

**Reply by NobodyXu** [reply]
> Overall Dioxus [tends to be faster than React](https://krausest.github.io/js-framework-benchmark/2024/table_chrome_123.0.6312.59.html), but I don't have benchmarks that specifically compare state management performance

Thanks, yes I do see that dioxus is faster than React, though its memory usage and generated wasm is quite large.

Also I didn't see dioxus 0.5 there, only found dioxus 0.4, so I can't compare dioxus 0.4 and 0.5 there.

> Targeting dioxus-web, the sync and unsync version should compile to similar code because atomics don't exist in wasm.

That's good to hear, so state management in web, when comparing with react, should have very low overhead.
